src/librustc_middle/traits/mod.rs

```diff
@@ -102,15 +135,29 @@ impl<'tcx> ObligationCause<'tcx> {
         body_id: hir::HirId,
         code: ObligationCauseCode<'tcx>,
     ) -> ObligationCause<'tcx> {
-        ObligationCause { span, body_id, code }
+        ObligationCause { data: Rc::new(ObligationCauseData { span, body_id, code }) }
     }
 
     pub fn misc(span: Span, body_id: hir::HirId) -> ObligationCause<'tcx> {
-        ObligationCause { span, body_id, code: MiscObligation }
+        ObligationCause::new(span, body_id, MiscObligation)
+    }
+
+    pub fn at_span(span: Span) -> ObligationCause<'tcx> {
+        ObligationCause::new(span, hir::CRATE_HIR_ID, MiscObligation)
     }
 
+    #[inline(always)]
     pub fn dummy() -> ObligationCause<'tcx> {
-        ObligationCause { span: DUMMY_SP, body_id: hir::CRATE_HIR_ID, code: MiscObligation }
+        DUMMY_OBLIGATION_CAUSE.with(|&x| x.clone())
+    }
+
+    #[inline(always)]
+    pub fn dummy_ref() -> &'static ObligationCause<'tcx> {
```

Comment:
Having a separate constructor just to avoid incrementing and decrementing a reference count is kind of a bummer. Is there reason to believe that this is actually worth doing? My guess is that this will be insignificant compared to the savings from making `ObligationCause` smaller. Can we try `dummy_ref` separately?

Reply:
Considering that this approach also causes a memory leak... yeah, let's just remove it.

I personally am a big fan of optimizing irrelevant things :sweat_smile: so I am kind of unhappy having to remove this